### PR TITLE
fix(gadget-sdk)!: prevent duplicate and self-referential messages

### PR DESCRIPTION
**Pull request by tbraun96**
This pull request includes several changes to improve the network and gossip handling in the `sdk` module. The updates mainly focus on enhancing the peer connection logic, adding new fields to structures, and refining the message handling process to avoid redundant operations and ensure efficient memory usage.

### Network and Gossip Handling Improvements:

* **Peer Connection Logic**:
  - Replaced `JoinSet` with `FuturesOrdered` to manage peer connection tasks more efficiently in `blueprint-manager/src/sdk/setup.rs`.
  - Added a new method `wait_for_peers` to handle peer connections asynchronously.

* **Structure Enhancements**:
  - Added `my_id` field to `NetworkServiceWithoutSwarm` and `NetworkService` structures to track the peer's own ID in `sdk/src/network/gossip.rs`. [[1]](diffhunk://#diff-eaa31da1831a49959d61586dde5384769520acbbe1476224c7abb52e4884e930R51) [[2]](diffhunk://#diff-eaa31da1831a49959d61586dde5384769520acbbe1476224c7abb52e4884e930R76)
  - Introduced `recent_messages` field using `LruCache` to store recently received message hashes and prevent processing duplicates in `GossipHandle`. [[1]](diffhunk://#diff-eaa31da1831a49959d61586dde5384769520acbbe1476224c7abb52e4884e930L250-R260) [[2]](diffhunk://#diff-02f8c36d33799a51af13963c7f32923063c25bef53943ffbed6ba16dc94f7955R272-R274)

* **Message Handling**:
  - Implemented logic to reject messages originating from the peer itself to avoid processing self-sent messages in `sdk/src/network/handlers/gossip.rs` and `sdk/src/network/handlers/p2p.rs`. [[1]](diffhunk://#diff-7257d41590a79cf8df1353111dbdd7612aa726471e37e80ad09c8921da05ab8dR81-R86) [[2]](diffhunk://#diff-118cb28d332ba4d6efe049f75450d0680aa50b427afaea8ed45fa68d9e388bc1R147-R151)
  - Improved message deserialization and added checks to prevent duplicate message processing using `recent_messages` in `GossipHandle`. [[1]](diffhunk://#diff-eaa31da1831a49959d61586dde5384769520acbbe1476224c7abb52e4884e930R345-R367) [[2]](diffhunk://#diff-eaa31da1831a49959d61586dde5384769520acbbe1476224c7abb52e4884e930R395-R405)

* **Test Improvements**:
  - Refactored test functions for clarity and removed unnecessary synchronization barriers in `sdk/src/network/mod.rs`. [[1]](diffhunk://#diff-c84d55c471bd61eec551e94967d187c1009a3ed86e5afa0a1ca8ea8a82671fd1L463-R462) [[2]](diffhunk://#diff-c84d55c471bd61eec551e94967d187c1009a3ed86e5afa0a1ca8ea8a82671fd1L490-R487) [[3]](diffhunk://#diff-c84d55c471bd61eec551e94967d187c1009a3ed86e5afa0a1ca8ea8a82671fd1L557-R555) [[4]](diffhunk://#diff-c84d55c471bd61eec551e94967d187c1009a3ed86e5afa0a1ca8ea8a82671fd1L665-L678)

* **Dependency Updates**:
  - Added `lru-mem` dependency to manage the LRU cache for recent messages in `Cargo.toml` and `sdk/Cargo.toml`. [[1]](diffhunk://#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R199) [[2]](diffhunk://#diff-0e065fd866dcb2fcbbae52037a4d993b2d3d3cee6b05a401066265b258e60773R95-R96)

### CODE REVIEW

File: blueprint-manager/src/sdk/setup.rs [hunk 1]
```diff
@@ -110,28 +111,31 @@ pub async fn wait_for_connection_to_bootnodes(
...
     }
+
     // Wait for all tasks to finish
-    while tasks.join_next().await.is_some() {}
+    tasks.collect::<()>().await;
 
     Ok(())
 }
+
+async fn wait_for_peers(handle: &GossipHandle, n_required: usize) {
```
Comment (drewstone) [inline blueprint-manager/src/sdk/setup.rs:129]:
```suggestion
async fn wait_for_peers(handle: &GossipHandle, required: usize) {
```

File: blueprint-manager/src/sdk/setup.rs [hunk 2]
```diff
@@ -110,28 +111,31 @@ pub async fn wait_for_connection_to_bootnodes(
...
+    loop {
+        let n_connected = handle.connected_peers();
+        if n_connected >= n_required {
```
Comment (drewstone) [inline blueprint-manager/src/sdk/setup.rs:132]:
```suggestion
        if n_connected >= required {
```

File: blueprint-manager/src/sdk/setup.rs [hunk 3]
```diff
@@ -110,28 +111,31 @@ pub async fn wait_for_connection_to_bootnodes(
...
+            return;
+        }
+        let topic = handle.topic();
+        debug!(
+            "`{topic}`: We currently have {n_connected}/{n_required} peers connected to network"
```
Comment (drewstone) [inline blueprint-manager/src/sdk/setup.rs:137]:
```suggestion
            "`{topic}`: We currently have {n_connected}/{required} peers connected to network"
```

**Review by drewstone — COMMENTED** [review]
Nit

**Review by shekohex — APPROVED** [review]

**Pull request merged**